File: 최경열/BFS/단지번호붙이기/house.js
const readline =require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});
let input=[]
rl.on("line",function(line){
    input.push(line)
}).on("close",function(){
    const n = Number(input[0])
    const graph=[]
    const arr=[]
    const dx = [0, 0, -1, 1]
    const dy = [-1, 1, 0, 0]

    for(let i=1; i<=n; i++){
        graph.push(input[i].split("").map(Number));
    }
    function bfs(x,y){
        const queue = []
        queue.push([x,y])
        graph[x][y]=0
        let count=1
        while(queue.length){
            const target =queue.shift()
            x=target[0]
            y=target[1]
            for(let i=0; i<4; i++){
                nx = x+ dx[i]
                ny = y+ dy[i]
                if(nx <= -1 || nx >= n || ny <= -1 || ny >= n){
                    continue
                }
                if(graph[nx][ny]==1){
                    graph[nx][ny] = 0
                    queue.push([nx, ny])
                    count += 1
                }
            }
        }
        return count
    }

    for(let i =0; i<n; i++){
        for(let j=0; j<n; j++){
            if(graph[i][j] == 1){
                arr.push(bfs(i, j))
            }
        }
    }
    
    arr.sort((a,b)=>a-b)
    console.log(arr.length)

    for(let i in arr){
        console.log(arr[i])
    }
})

File: 최경열/BFS/단지번호붙이기/house.py
from collections import deque

graph = []
arr = []
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

n = int(input())
for i in range(n):
    graph.append(list(map(int, input())))


def bfs(x, y):
    queue = deque()
    queue.append((x, y))
    graph[x][y] = 0
    count = 1
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx <= -1 or nx >= n or ny <= -1 or ny >= n:
                continue
            if graph[nx][ny] == 1:
                graph[nx][ny] = 0
                queue.append((nx, ny))
                count += 1
    return count


for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            arr.append(bfs(i, j))

arr.sort()
print(len(arr))
for i in range(len(arr)):
    print(arr[i])
